Studio-Lovelies/GG-JointJustice-Unity
# Sharded tests

## Summary

This PR automatically shards tests based on the directory structure.

Previously our 4 test locations (EditMode, PlayMode Scripts, PlayMode Scenes, PlayMode Playthrough) were hard-coded into the pipeline file.

As tests will take longer as we add more cases, they should be sharded based on the directories they're in, without requiring changes to the build + test pipeline files.

## Before/after screenshots and/or animated gif
To achieve this, the test directory structure has been updated and tests will now be sharded based on the following syntax:  
`/unity-ggjj/Assets/Tests/{type}ModeTests/Suites/{subdirectory}`
For each `{subdirectory}` inside the `Suites` directory a parallel-running `{type}Mode` test job is started.

![image](https://user-images.githubusercontent.com/1689033/206927021-b0025a82-9ed6-497f-bf73-bab1835a51d0.png)

Note: Tests were previously located inside `/unity-ggjj/Assets/Tests/{type}ModeTests`. The `Suites` subdirectory has been added and test suites have been moved there, as we had test tools for the PlayMode tests in the same directory as the actual test suites.  
The Unity Test Runner doesn't allow specifying which directory to test, so the pipeline instead removes all directories on the same level. Without this additional directory, this resulted in TestTools missing and compilation failing.

As an example: A run of `PlayModeTests/Suites/Scripts` simply [copies that directory away, deletes all other suites (`/Scenes` and `/Playthrough`), and copies `PlayModeTests/Suites/Scripts` back in place](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/c28b3e68de4ee4b11e0d16cc7ffa553651113336/.github/workflows/execute_tests.yml#L44-L47).

Additionally, the [`Build game`](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/c28b3e68de4ee4b11e0d16cc7ffa553651113336/.github/workflows/build_game.yml) and [`Execute tests`](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/c28b3e68de4ee4b11e0d16cc7ffa553651113336/.github/workflows/execute_tests.yml) snippets of the pipeline have been split into [reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows).

## Testing instructions

0. Fork this repository and push the `features/247-playmode-tests` of this instance to the `develop` branch of your fork (this is required, as you can't manually run Actions from inside branches, that aren't yet committed to the default branch)
    ![image](https://user-images.githubusercontent.com/1689033/206927137-3ac455aa-90c8-4791-86e0-c48263458e28.png)
2. [Queue a build+test run](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/actions/workflows/build_and_test.yml) for this branch
3. Wait for the run to complete successfully
4. Verify the test and coverage results inside GitHub and Codecov.io are correct

## Additional information

- `[ ]` Changes UI
- `[ ]` Introduces new feature
- `[ ]` Removes existing feature
- `[x]` Has associated resource: 
  - closes #247
